### ISSUE TITLE
Update in the way they propose to extract the codebooks' embeddings

### ISSUE THREAD

**Issue by pedrocg42** [lucidrains/discrete-key-value-bottleneck-pytorch#2]
In the last version of the [paper](https://arxiv.org/pdf/2207.11240.pdf) they propose to apply a random projection to the encoder embedding to extract the C embedding heads used to find the keys of each codebook, instead of taking a portion of the actual output of the encoder embedding. This changes a little the implementation, just for everybody to be aware. 
![image](https://user-images.githubusercontent.com/44102827/230486452-151c85fe-4593-4620-ad83-168e8c65bacc.png)

**Comment by lucidrains** [comment]
@pedrocg42 oh interesting, does this random projection improve their results even more?

**Comment by lucidrains** [comment]
@pedrocg42 https://github.com/lucidrains/discrete-key-value-bottleneck-pytorch/commit/24212c8c4ee8ddd08d8fccb5130756cf35357982

**Comment by pedrocg42** [comment]
Thank you so much for addressing the change!

To be honest I don't know why they do it, I think they do it in order to have more control on the input dimension to the codebooks, and not depend so much on wether the output dimension of the encoder is divisible or not. In the paper they mention "For tasks where we judge x to be sufficiently low-dimensional, we skip encoding and partition x directly", so it looks this is the reason, to reduce the dimensions of x.

The implementation looks good, but if I understand well what you did, I think in the paper it has one **different** random projection per codebook, I think we would use the same random projection for every embeddingin your implementation right?

They released a third version of the paper where they change a lot the whole vision on the DKVB. Now the decoder is not trainable for example to make it compatible with Class Incremental Learning classic benchmarks.

**Comment by pedrocg42** [comment]
I am pretty sure by what they explain and by the changes in the figure figure in the decoder part of the architecture, what they propose now is not to concatenate the splitted values after the DKVB, but to have C (number of classes) codebooks, retrieve C values (without concatenating them), and for the average pooling of each of those values to be the prediction applying a softmax directly to those outputs. 

Really hard to explain sorry. haha

**Comment by lucidrains** [comment]
@pedrocg42 thanks for the succinct explanation of their changes! i am usually unable to keep up with updated versions of papers, so this is helpful

do you want to see if https://github.com/lucidrains/discrete-key-value-bottleneck-pytorch/commit/1d4ed99416c7293e596a679701c7443341baeb8c looks a bit better?

**Comment by pedrocg42** [comment]
That definitely looks better. I was trying to do it myself but my version was going to need some changes in the VQ repo as well. The only thing that I thing it is necessary is to control what is the output dimension of the random projection (I only added the projection dim parameter):
```python
class DiscreteKeyValueBottleneck(nn.Module):
    def __init__(
        self,
        dim,
        projection_dim,
        *,
        num_memories,
        num_memory_codebooks=1,
        encoder=None,
        dim_memory=None,
        average_pool_memories=True,
        **kwargs
    ):
        super().__init__()
        self.encoder = encoder

        self.vq = VectorQuantize(
            dim=projection_dim * num_memory_codebooks,
            codebook_dim=projection_dim,
            codebook_size=num_memories,
            heads=num_memory_codebooks,
            separate_codebook_per_head=True,
            **kwargs
        )

        dim_memory = default(dim_memory, projection_dim)
        self.values = nn.Parameter(
            torch.randn(num_memory_codebooks, num_memories, dim_memory)
        )

        rand_proj = torch.empty(num_memory_codebooks, dim, projection_dim)
        nn.init.xavier_normal_(rand_proj)

        self.register_buffer("rand_proj", rand_proj)
        self.average_pool_memories = average_pool_memories
```
That works for me, the other way the projections were of the same shape as the embedding. I think. 

Thank so much for being so nice and update the code so fast!

**Comment by lucidrains** [comment]
@pedrocg42 yea no problem! i'm curious where this research goes

https://github.com/lucidrains/discrete-key-value-bottleneck-pytorch/commit/8c63060c147809eb72d1293b5ce6e37b5b0c9018 decided to have `dim` as the base dimension for the vq codebooks (and what is randomly projected to), and `dim_embed` set separately for the embedding dimension, if it differs from `dim`

**Comment by pedrocg42** [comment]
@lucidrains looks amazing! Thanks so much for all the changes and your work. I am going to close the issue! 🤗
